File: Graphs/LC_261_Graph_Valid_Tree.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.18

class Solution {
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        if (edges.size() != n-1){
            return false;
        } 
        vector<int> root;
        for (int i = 0; i < n; i++){
            root.push_back(i);
        }
        for (vector<int> edge: edges){
            if (!doUnion(edge[0], edge[1], root)){
                return false;
            }
        }
        return true;
    }
private:
    int find(int x, vector<int>& root){
        if (x != root[x]){
            root[x] = find(root[x], root);
        }
        return root[x];
    }

    bool doUnion(int x, int y, vector<int>& root){
        int rx = find(x, root), ry = find(y, root);
        if (rx != ry){
            root[ry] = rx;
            return true;
        }
        return false;
    }
};

File: Graphs/LC_261_Graph_Valid_Tree.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.18

class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        # union find based method
        root = [i for i in range(n)]
        self.total = n
        def find(x):
            if root[x] != x:
                root[x] = find(root[x])
            return root[x]
        
        def union(x, y):
            rx, ry = find(x), find(y)
            if rx != ry:
                root[ry] = rx
                self.total -= 1
                return True
            else:
                return False
        
        for v1, v2 in edges:
            if not union(v1, v2):
                return False
        return self.total == 1
        

        """
        mapping = collections.defaultdict(list)
        for v1, v2 in edges:
            mapping[v1].append(v2)
            mapping[v2].append(v1)
        seen = set()

        def dfs(start, prev):
            if start in seen:
                return False
            seen.add(start)
            for nxt in mapping[start]:
                if nxt == prev:
                    continue
                if not dfs(nxt, start):
                    return False
            return True
        
        
        return dfs(0, -1) and len(seen) == n
        """
